lib/ide/src/transport/web.rs

```diff
@@ -0,0 +1,250 @@
+            // anyway, nothing useful to pass to caller. Error code or reason
+            // string should not be relied upon.
+            tx_clone.unbounded_send(Err(())).ok();
+        });
+        self.set_on_open(move |_| {
+            tx.unbounded_send(Ok(())).ok();
+        });
+
+        match rx.next().await {
+            Some(Ok(())) => {
+                self.clear_callbacka();
+                Ok(())
+            }
+            _ => Err(ConnectingError::FailedToConnect)
+        }
+    }
+
+    /// Checks the current state of the connection.
+    pub fn state(&self) -> State {
+        State::query_ws(&self.ws)
+    }
+
+    /// Sets callback for the `close` event.
+    pub fn set_on_close(&mut self, f:impl FnMut(CloseEvent) + 'static) {
+        self.on_close.wrap(f);
+        self.ws.set_onclose(self.on_close.js_ref());
+    }
+
+    /// Sets callback for the `error` event.
+    pub fn set_on_error(&mut self, f:impl FnMut(Event) + 'static) {
+        self.on_error.wrap(f);
+        self.ws.set_onerror(self.on_error.js_ref());
+    }
+
+    /// Sets callback for the `message` event.
+    pub fn set_on_message(&mut self, f:impl FnMut(MessageEvent) + 'static) {
+        self.on_message.wrap(f);
+        self.ws.set_onmessage(self.on_message.js_ref());
+    }
+
+    /// Sets callback for the `open` event.
+    pub fn set_on_open(&mut self, f:impl FnMut(Event) + 'static) {
+        self.on_open.wrap(f);
+        self.ws.set_onopen(self.on_open.js_ref());
+    }
+
+    /// Clears all the available callbacks.
+    pub fn clear_callbacka(&mut self) {
+        self.on_close  .clear();
+        self.on_error  .clear();
+        self.on_message.clear();
+        self.on_open   .clear();
+        self.ws.set_onclose(None);
+        self.ws.set_onerror(None);
+        self.ws.set_onmessage(None);
+        self.ws.set_onopen(None);
+    }
+}
+
+impl Transport for WebSocket {
+    fn send_text(&mut self, message:String) -> Result<(), Error> {
+        // Sending through the closed WebSocket can return Ok() with error only
+        // appearing in the log. We explicitly check for this to get failure as
+        // early as possible.
+        //
+        // If WebSocket closes after the check, caller will be able to handle it
+        // when receiving `TransportEvent::Closed`.
+        let state = self.state();
+        if state != State::Open {
+            Err(SendingError::NotOpen(state).into())
+        } else {
+            self.ws.send_with_str(&message).map_err(|e| {
+                SendingError::FailedToSend(js_to_string(e)).into()
+            })
+        }
+    }
+
+    fn set_event_tx(&mut self, tx:mpsc::UnboundedSender<TransportEvent>) {
+        let tx1 = tx.clone();
+        self.set_on_message(move |e| {
+            let data = e.data();
+            if let Some(text) = data.as_string() {
+                let _ = tx1.unbounded_send(TransportEvent::TextMessage(text));
```

Comment:
I still see the `let _` and you said that you will make a function for "send without caring about no recipients"